go/store/nbs/table_reader.go

```diff
@@ -519,60 +519,69 @@ func (tr tableReader) getManyAtOffsetsWithReadFunc(
 		offsets offsetRecSlice,
 		stats *Stats) error,
 ) {
-	// Now |offsetRecords| contains all locations within the table which must be search (note
-	// that there may be duplicates of a particular location). Sort by offset and scan forward,
-	// grouping sequences of reads into large physical reads.
+	type readBatch struct {
+		batch     offsetRecSlice
+		readStart uint64
+		readEnd   uint64
+	}
+	batchCh := make(chan readBatch, 128)
+	buildBatches := func() {
+		defer close(batchCh)
 
-	var batch offsetRecSlice
-	var readStart, readEnd uint64
+		// |offsetRecords| contains all locations within the table
+		// which must be search in sorted order and without
+		// duplicates. Now scan forward, grouping sequences of reads
+		// into larger physical reads.
 
-	for i := 0; i < len(offsetRecords); {
-		if ae.IsSet() {
-			return
-		}
+		var batch offsetRecSlice
+		var readStart, readEnd uint64
 
-		rec := offsetRecords[i]
-		length := tr.lengths[rec.ordinal]
+		for i, rec := range offsetRecords {
+			if ae.IsSet() {
+				break
+			}
+			length := tr.lengths[rec.ordinal]
+
+			if batch == nil {
+				batch = make(offsetRecSlice, 1)
+				batch[0] = offsetRecords[i]
+				readStart = rec.offset
+				readEnd = readStart + uint64(length)
+				continue
+			}
 
-		if batch == nil {
-			batch = make(offsetRecSlice, 1)
-			batch[0] = offsetRecords[i]
-			readStart = rec.offset
-			readEnd = readStart + uint64(length)
-			i++
-			continue
-		}
+			if newReadEnd, canRead := canReadAhead(rec, tr.lengths[rec.ordinal], readStart, readEnd, tr.blockSize); canRead {
+				batch = append(batch, rec)
+				readEnd = newReadEnd
+				continue
+			}
 
-		if newReadEnd, canRead := canReadAhead(rec, tr.lengths[rec.ordinal], readStart, readEnd, tr.blockSize); canRead {
-			batch = append(batch, rec)
-			readEnd = newReadEnd
-			i++
-			continue
+			batchCh <- readBatch{batch, readStart, readEnd}
+			batch = nil
 		}
 
-		wg.Add(1)
-		goReadStart := readStart
-		goReadEnd := readEnd
-		goBatch := batch
-		go func(batch offsetRecSlice) {
-			defer wg.Done()
-			err := readAtOffsets(ctx, goReadStart, goReadEnd, reqs, goBatch, stats)
-			ae.SetIfError(err)
-		}(batch)
-		batch = nil
-	}
-
-	if !ae.IsSet() {
-		if batch != nil {
-			wg.Add(1)
-			go func(batch offsetRecSlice) {
-				defer wg.Done()
-				err := readAtOffsets(ctx, readStart, readEnd, reqs, batch, stats)
+		if !ae.IsSet() && batch != nil {
+			batchCh <- readBatch{batch, readStart, readEnd}
+		}
+	}
+	readBatches := func() {
+		defer wg.Done()
+		for rb := range batchCh {
+			if !ae.IsSet() {
+				err := readAtOffsets(ctx, rb.readStart, rb.readEnd, reqs, rb.batch, stats)
 				ae.SetIfError(err)
-			}(batch)
-			batch = nil
+			}
 		}
 	}
+
+	go buildBatches()
+
+	ioParallelism := 4
+
+	wg.Add(ioParallelism)
+	for i := 0; i < ioParallelism; i++ {
+		go readBatches()
```

Comment:
I like to see my wg.Done() calls close to my wg.Add calls.  Makes me feel like I'm not missing something.  So I would do:

    wg.Add(ioParallelism)
    for i := 0; i < ioParallelism; i++ {
        go func() {
            defer wg.Done()
            readBatches()
        }()
    }

Then remove the wg.Done from the top of your readBatches func.